group-coordinator/src/main/java/org/apache/kafka/coordinator/group/RecordHelpers.java

```diff
@@ -131,13 +131,21 @@ public static Record newGroupSubscriptionMetadataRecord(
         Map<String, TopicMetadata> newSubscriptionMetadata
     ) {
         ConsumerGroupPartitionMetadataValue value = new ConsumerGroupPartitionMetadataValue();
-        newSubscriptionMetadata.forEach((topicName, topicMetadata) ->
+        newSubscriptionMetadata.forEach((topicName, topicMetadata) -> {
+            ArrayList<ConsumerGroupPartitionMetadataValue.PartitionMetadata> partitionRacks = new ArrayList<>();
+            topicMetadata.partitionRacks().forEach((partition, racks) -> {
+                partitionRacks.add(new ConsumerGroupPartitionMetadataValue.PartitionMetadata()
+                    .setPartition(partition)
+                    .setRacks(new ArrayList<>(racks))
```

Comment:
Do we set racks all the time or only when the set is non-empty?

Reply:
I changed the code now, if no racks are available then an empty map is stored in the topicMetadata and therefore in the subscription metadata. Moreover, an empty list will be stored in the record to save memory